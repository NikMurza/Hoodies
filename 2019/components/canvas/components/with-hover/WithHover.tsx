import * as React from 'react';
import Konva from 'konva';
import {
  VisualEditorActions,
  VisualEditorShapeType,
  VisualEditorShape,
} from '@common/components/lib/types/visual-editor';

export interface WithHoverProps {
  id: string;
  type: VisualEditorShapeType;
  onAction<T extends object>(action: VisualEditorActions, data: T): void;
  onMouseEnter?(e: Konva.KonvaEventObject<MouseEvent>): void;
  onMouseLeave?(e: Konva.KonvaEventObject<MouseEvent>): void;
}

export const WithHover = <P extends object>(
  WrappedComponent: React.ComponentType<P>,
): React.FC<P & WithHoverProps> => (props: WithHoverProps): JSX.Element => {
  const { id, type, onAction, onMouseEnter, onMouseLeave } = props;

  const triggerAction = React.useCallback(
    (e: Konva.KonvaEventObject<MouseEvent>, isOut: boolean): void => {
      const data = { id: isOut ? '' : id, type };
      onAction<VisualEditorShape>(VisualEditorActions.Hover, data);
    },
    [onAction],
  );

  const onLeave = React.useCallback(
    (e: Konva.KonvaEventObject<MouseEvent>): void => {
      triggerAction(e, true);
      if (onMouseLeave) {
        onMouseLeave(e);
      }
    },
    [triggerAction, onMouseLeave],
  );

  const onEnter = React.useCallback(
    (e: Konva.KonvaEventObject<MouseEvent>): void => {
      triggerAction(e, false);
      if (onMouseEnter) {
        onMouseEnter(e);
      }
    },
    [triggerAction, onMouseEnter],
  );

  return <WrappedComponent {...(props as P)} onMouseEnter={onEnter} onMouseLeave={onLeave} />;
};

export default WithHover;
