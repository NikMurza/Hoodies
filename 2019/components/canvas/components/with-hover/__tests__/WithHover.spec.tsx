import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import {
  VisualEditorActions,
  VisualEditorShapeType,
} from '@common/components/lib/types/visual-editor';
import { WithHover, WithHoverProps } from '../WithHover';

const onMouseEnter = jest.fn();
const onMouseLeave = jest.fn();

const props: WithHoverProps = {
  id: 'id1',
  type: VisualEditorShapeType.Circle,
  onAction: jest.fn,
  onMouseEnter,
  onMouseLeave,
};

describe('WithHover HOC', () => {
  it('renders correctly', () => {
    const HOC = WithHover(() => <div />);
    const wrapper = shallow(<HOC {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('works correctly', () => {
    const onActionSpy = spyOn(props, 'onAction');
    const HOC = WithHover(() => <div />);
    const wrapper = shallow(<HOC {...props} />);

    wrapper.simulate('mouseleave');
    expect(onMouseLeave).toHaveBeenCalled();
    expect(onActionSpy).toBeCalledWith(VisualEditorActions.Hover, {
      id: '',
      type: VisualEditorShapeType.Circle,
    });

    wrapper.simulate('mouseenter');
    expect(onMouseEnter).toHaveBeenCalled();
    expect(onActionSpy).toBeCalledWith(VisualEditorActions.Hover, {
      id: 'id1',
      type: VisualEditorShapeType.Circle,
    });
  });
});
