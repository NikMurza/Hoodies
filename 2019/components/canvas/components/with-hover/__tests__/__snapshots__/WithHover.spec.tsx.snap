// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithHover HOC renders correctly 1`] = `
<Component
  id="id1"
  onAction={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="circle"
/>
`;
